OpenCv3, OCL :  Assertion failed clEnqueueUnmapMemObject_pfn

Transferred from http://code.opencv.org/issues/4277

```
|| Thibault Lelore on 2015-04-13 17:57
|| Priority: Low
|| Affected: branch 'master' (3.0-dev)
|| Category: t-api
|| Tracker: Bug
|| Difficulty: Medium
|| PR: 
|| Platform: x86 / Windows
```
## OpenCv3, OCL :  Assertion failed clEnqueueUnmapMemObject_pfn

```
Hi!
I'm using latest version of opencv (3.0-dev git), and sometime my code throws an exception:
@OpenCV Error: Assertion failed (clEnqueueUnmapMemObject_pfn(q, (cl_mem)u->handle, data, 0, 0, 0) == 0) in cv::ocl::OpenCLAllocator::deallocate, file D:\XXX\modules\core\src\ocl.cpp, line 4446
exception caught: D:\XXX\modules\core\src\ocl.cpp:4446: error: (-215) clEnqueueUnmapMemObject_pfn(q, (cl_mem)u->handle, data, 0, 0, 0) == 0 in function cv::ocl::OpenCLAllocator::deallocate@

It seems to be related to ORB (OCL version)... I first compute points, then descriptors, and after 50 iterations, the exception is thrown!
```
## History
##### Thibault Lelore on 2015-04-14 10:10

```
I'm sorry, the problem was my side (memory corruption).
-   Status changed from New to Cancelled
```
##### Thibault Lelore on 2015-04-14 13:45

```
Finally, I think there is a problem in ORB implementation... Here is the smallest code I use which produce the previous error: Assertion failed (clEnqueueUnmapMemObject_pfn...
<pre>
VideoCapture cap(0);
  cv::Ptr<cv::xfeatures2d::SURF> ptDetect = xfeatures2d::SURF::create();
  cv::Ptr<cv::ORB> descExtract = ORB::create();
  cv::Mat frame, desc;
  std::vector<cv::KeyPoint> points;
  while (1)
  {
    cap >> frame; // read a new frame from video

    if (frame.empty()) //if not success, break loop
      break;

    ptDetect->detect(frame, points);
    descExtract->compute(frame, points, desc);

    imshow("MyVideo", frame);
    waitKey(15);
  }
</pre>
After 15-30 frames, the exception is thrown... Can someone confirm this issue?
-   Status changed from Cancelled to Open
```
##### Maksim Shabunin on 2015-04-27 10:50

```
Can not reproduce with your code on Linux, Intel HD Graphics Haswell GT2, using beignet 1.0.0.
Can you, please, provide some information about your platform
-   Target version changed from 3.0 to 3.1
-   Priority changed from Normal to Low
-   Assignee set to Maksim Shabunin
-   Status changed from Open to Incomplete
-   Category set to ocl
```
##### Vadim Pisarevsky on 2015-04-28 12:05

```
-   Category changed from ocl to t-api
```

Same error
On windows 7 64-bit
Opencv 3.0.0 32-bit static-lib
Visual Studio  2010

..\..\..\modules\core\src\ocl.cpp:4446: error: (-215) clEnqueueUnmapMemObject_pfn(q, (cl_mem)u->handle, data, 0, 0, 0) == 0 in function cv::ocl::OpenCLAllocator::deallocate

--- a/modules/core/include/opencv2/core/ocl.hpp
+++ b/modules/core/include/opencv2/core/ocl.hpp
@@ -739,6 +739,9 @@ CV_EXPORTS MatAllocator* getOpenCLAllocator();
 #ifdef __OPENCV_BUILD
 namespace internal {
 
+CV_EXPORTS bool isOpenCLForced();
+#define OCL_FORCE_CHECK(condition) (cv::ocl::internal::isOpenCLForced() || (condition))
+
 CV_EXPORTS bool isPerformanceCheckBypassed();
 #define OCL_PERFORMANCE_CHECK(condition) (cv::ocl::internal::isPerformanceCheckBypassed() || (condition))
 

--- a/modules/core/src/ocl.cpp
+++ b/modules/core/src/ocl.cpp
@@ -6073,6 +6073,18 @@ void* Image2D::ptr() const
     return p ? p->handle : 0;
 }
 
+bool internal::isOpenCLForced()
+{
+    static bool initialized = false;
+    static bool value = false;
+    if (!initialized)
+    {
+        value = getBoolParameter("OPENCV_OPENCL_FORCE", false);
+        initialized = true;
+    }
+    return value;
+}
+
 bool internal::isPerformanceCheckBypassed()
 {
     static bool initialized = false;

--- a/modules/features2d/src/orb.cpp
+++ b/modules/features2d/src/orb.cpp
@@ -972,7 +972,7 @@ void ORB_Impl::detectAndCompute( InputArray _image, InputArray _mask,
     int halfPatchSize = patchSize / 2;
     int border = std::max(edgeThreshold, std::max(halfPatchSize, HARRIS_BLOCK_SIZE/2))+1;
 
-    bool useOCL = ocl::useOpenCL();
+    bool useOCL = ocl::useOpenCL() && OCL_FORCE_CHECK(_image.isUMat() || _descriptors.isUMat());
 
     Mat image = _image.getMat(), mask = _mask.getMat();
     if( image.type() != CV_8UC1 )

--- a/modules/objdetect/src/cascadedetect.cpp
+++ b/modules/objdetect/src/cascadedetect.cpp
@@ -1290,8 +1290,8 @@ void CascadeClassifierImpl::detectMultiScaleNoGrouping( InputArray _image, std::
 
 #ifdef HAVE_OPENCL
     bool use_ocl = tryOpenCL && ocl::useOpenCL() &&
+         OCL_FORCE_CHECK(_image.isUMat()) &&
          featureEvaluator->getLocalSize().area() > 0 &&
-         ocl::Device::getDefault().type() != ocl::Device::TYPE_CPU &&
          (data.minNodesPerTree == data.maxNodesPerTree) &&
          !isOldFormatCascade() &&
          maskGenerator.empty() &&
@@ -1316,7 +1316,8 @@ void CascadeClassifierImpl::detectMultiScaleNoGrouping( InputArray _image, std::
     // OpenCL code
     CV_OCL_RUN(use_ocl, ocl_detectMultiScaleNoGrouping( scales, candidates ))
 
-    tryOpenCL = false;
+    if (use_ocl)
+        tryOpenCL = false;
 #endif
 
     // CPU code

